src/workerd/api/memory-cache.h

```diff
@@ -0,0 +1,512 @@
+// within the same process.
+// TODO(soon): We plan to explore replacing this implementation with a memcached-based
+// implementation in the near future. The memcached-based impl would likely be
+// fairly different from this implementation so quite a few of the details here
+// are expected to change.
+class SharedMemoryCache : public kj::AtomicRefcounted {
+private:
+  struct InProgress;
+
+public:
+  struct ThreadUnsafeData;
+
+  struct Limits {
+    // The maximum number of keys that may exist within the cache at the same
+    // time. The cache size grows at least linearly in the number of entries.
+    uint32_t maxKeys;
+
+    // The maximum size of each individual value, when serialized.
+    uint32_t maxValueSize;
+
+    // The maximum sum of all stored values. This is essentially the cache size,
+    // except that it only includes the sizes of the values and does not account
+    // for keys and the overhead of the data structures themselves.
+    uint64_t maxTotalValueSize;
+
+    bool operator<(const Limits& b) const {
+      if (maxTotalValueSize != b.maxTotalValueSize) {
+        return maxTotalValueSize < b.maxTotalValueSize;
+      }
+      if (maxKeys != b.maxKeys) {
+        return maxKeys < b.maxKeys;
+      }
+      return maxTotalValueSize < b.maxTotalValueSize;
+    }
+
+    Limits normalize() const KJ_WARN_UNUSED_RESULT {
+      // Avoid surprises due to misconfigured bindings that set one or more limits to 0.
+      if (maxKeys == 0 || maxValueSize == 0 || maxTotalValueSize == 0) {
+        return min();
+      }
+
+      // If a binding specifies a maxValueSize that exceeds the maxTotalValueSize, remedy
+      // that by reducing the maxValueSize.
+      return Limits{
+        .maxKeys = maxKeys,
+        .maxValueSize = static_cast<uint32_t>(kj::min(maxValueSize, maxTotalValueSize)),
+        .maxTotalValueSize = maxTotalValueSize,
+      };
+    }
+
+    static constexpr Limits min() { return {0, 0, 0}; }
+
+    static Limits max(const Limits& a, const Limits& b) {
+      return Limits{
+        std::max(a.maxKeys, b.maxKeys),
+        std::max(a.maxValueSize, b.maxValueSize),
+        std::max(a.maxTotalValueSize, b.maxTotalValueSize),
+      };
+    }
+  };
+
+  KJ_DISALLOW_COPY_AND_MOVE(SharedMemoryCache);
+
+  using AdditionalResizeMemoryLimitHandler = kj::Function<void(ThreadUnsafeData&)>;
+
+  SharedMemoryCache(
+      kj::Maybe<const MemoryCacheProvider&> provider,
+      kj::StringPtr id,
+      kj::Maybe<AdditionalResizeMemoryLimitHandler&> additionalResizeMemoryLimitHandler);
+
+  ~SharedMemoryCache() noexcept(false);
+
+  kj::StringPtr getId() const { return id; }
+
+  static kj::Own<const SharedMemoryCache> create(
+    kj::Maybe<const MemoryCacheProvider&> provider,
+    kj::StringPtr id,
+    kj::Maybe<AdditionalResizeMemoryLimitHandler&> additionalResizeMemoryLimitHandler);
+
+  kj::Maybe<kj::Own<const SharedMemoryCache>> tryAddRef() const;
```

Comment:
Why not just inline this into the one call site?

It seems like providing `tryAddRef()` as a helper method isn't that useful, because the only place that would want to use the `tryAddRef()` version is the case where were trying to upgrade the weak reference in the map to a strong ref. All other places that want to addRef() a SharedMemoryCache will already have a strong ref and so won't need or want the "try" part.